op-wheel/cheat/cheat.go

```diff
@@ -0,0 +1,340 @@
+package cheat
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"math/big"
+	"path/filepath"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/consensus/beacon"
+	"github.com/ethereum/go-ethereum/consensus/ethash"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/trie"
+)
+
+var HundredETH = big.NewInt(0).Mul(big.NewInt(100), big.NewInt(1000000000000000000))
+
+type Cheater struct {
+	// The database of the chain with the head block that we patch the state-root of, once the state is updated.
+	DB ethdb.Database
+	// Initialized chain, wrapping the database with in-memory presentation of headers and recent changes and such.
+	Blockchain *core.BlockChain
+	// The Cheater avoids making writes if this is set to True, and opens the DB as readonly.
+	ReadOnly bool
+}
+
+func OpenGethRawDB(dataDirPath string, readOnly bool) (ethdb.Database, error) {
+	// don't use readonly mode in actual DB, it doesn't work with Geth.
+	db, err := rawdb.NewLevelDBDatabaseWithFreezer(dataDirPath, 2048, 500, filepath.Join(dataDirPath, "ancient"), "", readOnly)
```

Comment:
That's a pretty high number of db handles and I think it's over the default limit on mac osx. I also don't think you get much more performance by having so many.

Reply:
Perhaps using the default values of what geth uses makes sense

Reply:
Geth uses `MakeDatabaseHandles` to actually request more handles.

Reply:
I found the 500 as default somewhere in geth, but in other places it's dynamic yes. I'll see if I can adjust it, or make geth pick the number.

Reply:
On macs you can raise the limit to almost anything IIRC, but the default limit is like 256